Add "Stock" game mode; add mechanism to start and stop a round of the game, crowning a winner

cspotcode opened the pull request:
## The Big Idea

When a round of the game starts, a single `RoundController` and a single subclass of `GameMode` are created.

They work together to:

- start the round (`RoundController` does this)
- track kills or other win condition (`GameMode` does this)
- decide when we have a winner (`GameMode` does this)
- trigger victory animations, camerawork, splash text, etc (`RoundController` does this)

This PR includes `StockGameMode`, a subclass / implementation of the `GameMode` interface for a stock battle mode: last person standing, each player gets 3 lives.

## Diagram

Note that many things on this diagram already exist in the game today.

### Already exist

- Hud
- PlayerHud
- Player
- Vehicle

### Added by this PR

- RoundController
- StockGameMode
- StockGameModePlayerState

### Will be added by Hammerz

- LivesHud

```mermaid
classDiagram

note "NOTE:
This diagram depicts
Stock-specific subclasses of
of `GameMode` and `GameModePlayerState`.
"

StockGameMode <.. Player : Notifies on death

StockGameMode --> "*" Player : (Un)locks respawning
RoundController --> "*" Player : Resets to initial spawn on round start\n(Un)locks movement on "Go!"
RoundController --> StockGameMode : Queries for winner every frame

Player -- Vehicle

Player *-- StockGameModePlayerState

Hud o-- "*" PlayerHud
PlayerHud --> Player
PlayerHud -- StockGameModePlayerState
PlayerHud o-- LivesHud
LivesHud --> StockGameModePlayerState


class PlayerHud {
    Shows player's health, icon, & respawn timer.
}

class LivesHud {
    Shows heart icons for lives
}
note for LivesHud "This class will be added by Hammerz"

class RoundController {
    Shows 3-2-1 countdown
    Show "winner!" banner
    Triggers reset for next round start
}

class StockGameModePlayerState {
    Created and attached to each Player by StockGameMode
    ---
    +int lives
}

class Player {
    +bool controls_active
    +bool allowed_to_respawn
}
```

## Update logic

I rewrote the `update` logic to be a plain `update()` function so that no one would have to learn how generators work.  To compare against the older, generator implementation, see here.

Generator on the left, plain `update()` on the right:
https://github.com/CastIronChat/twisted-metal/pull/79/commits/59abae8603953be922d71c2563b9cd8267c76a36

cspotcode commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Prototype round lifecycle: stock](https://trello.com/c/0UidkCgz/81-prototype-round-lifecycle-stock)

cspotcode commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Use all spawn points for respawn, not just the initial spawn locations](https://trello.com/c/mOxLbKRj/153-use-all-spawn-points-for-respawn-not-just-the-initial-spawn-locations)

Thorakii commented on src/rounds/game_modes/stock.py:
```diff
@@ -0,0 +1,90 @@
+        if len(players_with_remaining_lives) == 0:
+            raise Exception(
+                "Everyone died at once; this is not implemented yet.  Should be declared a draw?"
+            )
```
Maybe the last players to die simultaneously get 1 life back for sudden death mode? Just a thought

cspotcode replied:
Ooh good idea.

I looked at this logic again, and I think it's actually impossible for two players to die simultaneously, because `on_player_death` is called sequentially per player that's dying, so it's guaranteed that one player will die before the other even if it's all happening in the same frame.

But I could add some sort of death grace period, where if you kill someone but then die immediately after -- within 0.2 seconds? -- then you both resurrect into sudden death.  Not sure if that's a fun or a terrible idea.

I made a Trello card for it.

Thorakii replied:
Yeah, I think this will be a good future branch. Either both players get an extra life to duke it out, or it says they tie.

From what you're describing, as is, If you get the final kill with a rocket explosion that hits you too, it will just give the win to whoever it checks first/last, giving an advantage to player order

Thorakii reviewed (approved):
Looks good and runs. I left a few comments and questions, but nothing blocking.

cspotcode commented on src/player.py:
```diff
@@ -1,43 +1,77 @@
+        self._spawn_points = spawn_points
         self.alive: bool = True
+        self.controls_active: bool = False
+        "Is false between rounds, when you're not allowed to move yet"
```
I thought it was clearer to track the two concepts separately, and there's logic in `def update` that checks `not alive` and updates a respawn timer.  So I didn't want respawn timer logic to be running during the 3-2-1 countdown.  I was also thinking future code might check `if not alive` to control an animation on the flame sprite, or something like that.

Thorakii replied:
That makes sense. Agreed

cspotcode commented on src/player.py:
```diff
@@ -1,43 +1,77 @@
+
+    round_controller: RoundController
+    "Necessary for player to notify the round controller and game mode about relevant events"
+
```
Hmm, yeah that's confusing.  It doesn't need access to `round_controller`, but it needs `game_mode`.  And `vehicle` is the one using it, since it's the thing that makes death happen, and it needs to notify the gamemode.

https://github.com/CastIronChat/twisted-metal/blob/59abae8603953be922d71c2563b9cd8267c76a36/src/vehicle.py#L96

I tweaked it to store `game_mode` instead of `round_controller`, to be a bit clearer.

Thorakii replied:
Ah, that's were it's called. Yeah, storing game_mode seems cleaner.

cspotcode commented on src/main.py:
```diff
@@ -54,6 +56,15 @@ def setup(self):
+
+        # Chicken-and-egg wiring
+        for player in self.player_manager.players:
+            player.round_controller = self.round_controller
```
Yeah, I'm guessing doing it here is a holdover from a previous iteration of this code.  I moved it into `RoundController` `__init__`

cspotcode commented on src/rounds/game_modes/stock.py:
```diff
@@ -0,0 +1,90 @@
+
+    def on_round_init(
+        self, players: list[Player], arena: Arena, sprite_lists: SpriteLists
+    ):
```
You're right, it doesn't.  This is a holdover from a previous iteration of the design.

cspotcode commented:
Thanks, I've addressed your feedback.

I also removed a couple lines of debug logging that I accidentally left in.

And I removed two `GameMode` methods `create_hud` and `create_player_hud` because Mike D and I
ended up tweaking how hud elements get created so those methods won't be needed.

Thorakii reviewed (approved):
Approved. Comments addressed, and branch runs
